File: Sources/SpeechX/TypeOverWatcher.swift
import Foundation
import ApplicationServices

/// Pure, testable logic for spotting a "type-over" spelling correction: a word
/// SpeechX injected that the user then edited in place into a close variant.
enum TypeOverDetector {

    /// Common words we never auto-learn — biasing Deepgram toward them (or toward
    /// one side of a homophone like there/their) would hurt recognition. Focus
    /// words are meant for names / jargon / out-of-vocabulary terms.
    static let commonWords: Set<String> = [
        "the","and","that","have","for","not","with","you","this","but","his","from","they",
        "say","her","she","will","one","all","would","there","their","they're","what","out","about",
        "who","get","which","when","make","can","like","time","just","him","know","take","into","your",
        "you're","year","good","some","could","them","see","other","than","then","now","look","only",
        "come","its","it's","over","think","also","back","after","use","two","too","to","how","our",
        "work","first","well","way","even","new","want","because","any","these","give","day","most",
        "here","hear","were","where","we're","been","being","are","our","off","of","on","in","is","as",
    ]

    /// Detect a single-word in-place correction. `injectedWords` are the (lowercased)
    /// words SpeechX just typed; `baseline` is the field right after injection;
    /// `current` is the field after the user edited it.
    static func correction(injectedWords: Set<String>, baseline: String, current: String) -> (original: String, corrected: String)? {
        let baseTokens = tokenize(baseline)
        let curTokens  = tokenize(current)

        // Count-based diff: how each word's occurrence count changed. Using counts
        // (not set membership) means it still works when a word repeats in the field
        // or when the corrected spelling already appears elsewhere.
        var delta: [String: Int] = [:]
        for w in baseTokens { delta[w.lowercased(), default: 0] -= 1 }
        for w in curTokens  { delta[w.lowercased(), default: 0] += 1 }

        let gained = delta.filter { $0.value > 0 }   // appears more now → the new spelling
        let lost   = delta.filter { $0.value < 0 }   // appears less now → the old spelling
        // Exactly one word swapped for exactly one other (each by a single occurrence).
        guard gained.count == 1, lost.count == 1,
              let added = gained.first, added.value == 1,
              let removed = lost.first, removed.value == -1 else { return nil }

        let originalLC = removed.key
        let correctedLC = added.key
        // The word that lost an occurrence must be one SpeechX injected.
        guard injectedWords.contains(originalLC) else { return nil }

        // The change must be a spelling variant, not an unrelated word…
        let dist = levenshtein(originalLC, correctedLC)
        let maxLen = max(originalLC.count, correctedLC.count)
        guard maxLen >= 3, dist >= 1, dist <= max(2, maxLen / 3) else { return nil }
        // …and not an everyday word.
        guard !commonWords.contains(correctedLC) else { return nil }

        // Preserve the words' real casing as they appear in the text.
        let original  = baseTokens.first { $0.lowercased() == originalLC } ?? originalLC
        let corrected = curTokens.first  { $0.lowercased() == correctedLC } ?? correctedLC
        return (original, corrected)
    }

    static func tokenize(_ s: String) -> [String] {
        s.components(separatedBy: CharacterSet.letters.union(CharacterSet(charactersIn: "'")).inverted)
            .filter { $0.contains(where: { $0.isLetter }) }
    }

    static func levenshtein(_ a: String, _ b: String) -> Int {
        let a = Array(a), b = Array(b)
        if a.isEmpty { return b.count }
        if b.isEmpty { return a.count }
        var prev = Array(0...b.count)
        var cur = [Int](repeating: 0, count: b.count + 1)
        for i in 1...a.count {
            cur[0] = i
            for j in 1...b.count {
                let cost = a[i-1] == b[j-1] ? 0 : 1
                cur[j] = Swift.min(prev[j] + 1, cur[j-1] + 1, prev[j-1] + cost)
            }
            swap(&prev, &cur)
        }
        return prev[b.count]
    }
}

/// C callback bridge — recovers the watcher from the refcon (it can't capture context).
private let axValueChangedCallback: AXObserverCallback = { _, _, _, refcon in
    guard let refcon else { return }
    Unmanaged<TypeOverWatcher>.fromOpaque(refcon).takeUnretainedValue().valueChanged()
}

/// Watches the focused text field (via Accessibility) for a short window after
/// SpeechX injects text, and reports an in-place spelling correction of an
/// injected word. Privacy: it only diffs against what was injected and keeps just
/// the wrong→right word pair — it never stores the surrounding field text.
final class TypeOverWatcher {

    /// Called on the main thread once the user has *finished* editing a word into
    /// its corrected spelling. `original` is the word that was typed and corrected
    /// away from, so the caller can drop it and its near-variants from the
    /// dictionary — collapsing successive re-spellings of the same word.
    var onCorrection: ((_ corrected: String, _ original: String) -> Void)?

    private var observer: AXObserver?
    private var element: AXUIElement?
    private var baseline = ""
    private var injectedWords: Set<String> = []
    private var stopWork: DispatchWorkItem?
    private var debounceWork: DispatchWorkItem?
    /// The word we've already auto-added this session — so further edits update it.
    private var lastLearned: String?
    private let watchSeconds: TimeInterval = 120
    /// How long the field must be idle before we treat an edit as finished.
    private let settleSeconds: TimeInterval = 2.0

    /// Begin watching after `injected` was typed into the focused field. The
    /// baseline snapshot is deferred briefly so the paste has settled.
    func watch(injected: String) {
        let words = Set(TypeOverDetector.tokenize(injected).map { $0.lowercased() })
        guard !words.isEmpty else { return }
        DispatchQueue.main.asyncAfter(deadline: .now() + 0.5) { [weak self] in
            self?.start(injectedWords: words)
        }
    }

    private func start(injectedWords words: Set<String>) {
        stop()
        guard AXIsProcessTrusted() else { return }

        let systemWide = AXUIElementCreateSystemWide()
        var focusedRef: CFTypeRef?
        guard AXUIElementCopyAttributeValue(systemWide, kAXFocusedUIElementAttribute as CFString, &focusedRef) == .success,
              let focused = focusedRef else { return }
        let el = focused as! AXUIElement

        guard let value = stringValue(of: el) else { return }   // field must expose its text
        var pid: pid_t = 0
        guard AXUIElementGetPid(el, &pid) == .success else { return }

        var obs: AXObserver?
        guard AXObserverCreate(pid, axValueChangedCallback, &obs) == .success, let obs else { return }
        let refcon = Unmanaged.passUnretained(self).toOpaque()
        AXObserverAddNotification(obs, el, kAXValueChangedNotification as CFString, refcon)
        CFRunLoopAddSource(CFRunLoopGetCurrent(), AXObserverGetRunLoopSource(obs), .defaultMode)

        self.observer = obs
        self.element = el
        self.baseline = value
        self.injectedWords = words

        let work = DispatchWorkItem { [weak self] in self?.stop() }
        stopWork = work
        DispatchQueue.main.asyncAfter(deadline: .now() + watchSeconds, execute: work)
    }

    /// Fires on every keystroke — we don't act yet. Instead we (re)arm a settle
    /// timer, so we only evaluate once the user pauses. Each new keystroke pushes
    /// the timer out, so mid-word states like "Aysh" never get learned.
    fileprivate func valueChanged() {
        debounceWork?.cancel()
        let work = DispatchWorkItem { [weak self] in self?.evaluate() }
        debounceWork = work
        DispatchQueue.main.asyncAfter(deadline: .now() + settleSeconds, execute: work)
    }

    /// Run after editing settles: learn the *final* spelling. If we already learned
    /// a (now-superseded) word this session, ask the caller to replace it.
    private func evaluate() {
        guard let el = element, let current = stringValue(of: el) else { return }
        guard let (original, corrected) = TypeOverDetector.correction(
            injectedWords: injectedWords, baseline: baseline, current: current
        ) else { return }
        guard corrected.lowercased() != lastLearned?.lowercased() else { return }
        lastLearned = corrected
        onCorrection?(corrected, original)
    }

    func stop() {
        stopWork?.cancel(); stopWork = nil
        debounceWork?.cancel(); debounceWork = nil
        if let obs = observer, let el = element {
            AXObserverRemoveNotification(obs, el, kAXValueChangedNotification as CFString)
            CFRunLoopRemoveSource(CFRunLoopGetCurrent(), AXObserverGetRunLoopSource(obs), .defaultMode)
        }
        observer = nil
        element = nil
        baseline = ""
        injectedWords = []
        lastLearned = nil
    }

    private func stringValue(of element: AXUIElement) -> String? {
        var ref: CFTypeRef?
        guard AXUIElementCopyAttributeValue(element, kAXValueAttribute as CFString, &ref) == .success else { return nil }
        return ref as? String
    }
}

File: windows/SpeechX/Core/TypeOverWatcher.cs
using System.Diagnostics;
using System.Windows.Automation;
using Timer = System.Threading.Timer;

namespace SpeechX.Core;

/// <summary>
/// Watches the focused text field (via UI Automation) for a short window after SpeechX injects
/// text, and reports an in-place spelling correction of an injected word. Windows port of the macOS
/// <c>TypeOverWatcher</c>: UIA's <c>AutomationElement.FocusedElement</c> stands in for the AX
/// <c>kAXFocusedUIElement</c>.
///
/// Change detection is by <b>polling</b> the element's text on a background timer rather than by UIA
/// change events: value/text-changed events are raised inconsistently across apps (classic Win32 edit
/// controls in particular), whereas re-reading the value — which we already confirmed is readable at
/// <see cref="Start"/> — works everywhere the field exposes its text. This mirrors the macOS AX
/// observer's effect without depending on any app to raise events correctly.
///
/// Privacy: it only diffs against what was injected and keeps just the wrong→right word pair — it
/// never stores the surrounding field text.
///
/// All UI Automation work runs on thread-pool threads (never the WPF UI thread), which is the
/// Microsoft-recommended pattern for UIA clients. Standard Win32 edit fields expose their text well;
/// browsers/Electron/terminals do so inconsistently (a platform limitation, same as the macOS AX story).
/// </summary>
public sealed class TypeOverWatcher
{
    /// <summary>
    /// Called once the user has *finished* editing a word into its corrected spelling. <c>original</c>
    /// is the word that was typed and corrected away from, so the caller can drop it and its
    /// near-variants from the dictionary — collapsing successive re-spellings of the same word.
    /// Invoked on a background thread; the handler is responsible for marshalling to the UI thread.
    /// </summary>
    public Action<string /* corrected */, string /* original */>? OnCorrection;

    private readonly object _gate = new();

    private AutomationElement? _element;
    private string _baseline = "";
    private string _lastPolled = "";
    private HashSet<string> _injectedWords = new();
    private string? _lastLearned;   // the word we've already auto-added this session

    private Timer? _startTimer;
    private Timer? _pollTimer;
    private Timer? _settleTimer;
    private Timer? _stopTimer;
    private int _polling;           // re-entrancy guard so a slow UIA read can't overlap the next tick

    private const int StartDelayMs = 500;      // let the paste settle before snapshotting the baseline
    private const int BaselineTimeoutMs = 3000; // keep re-reading up to this long for the paste to land
    private const int BaselinePollMs = 150;    // re-check cadence while waiting for the injected text
    private const int PollMs = 400;            // how often we re-read the field for edits
    private const int SettleMs = 2000;         // idle time before we treat an edit as finished
    private const int WatchMs = 120_000;       // stop watching after two minutes

    /// <summary>
    /// Begin watching after <paramref name="injected"/> was typed into the focused field. The baseline
    /// snapshot is deferred briefly so the paste has settled.
    /// </summary>
    public void Watch(string injected)
    {
        var words = TypeOverDetector.Tokenize(injected).Select(w => w.ToLowerInvariant()).ToHashSet();
        if (words.Count == 0) return;

        lock (_gate)
        {
            _startTimer?.Dispose();
            _startTimer = new Timer(_ => Start(words), null, StartDelayMs, Timeout.Infinite);
        }
    }

    private void Start(HashSet<string> words)
    {
        Stop();
        try
        {
            // Snapshot the baseline only once the paste has actually landed. We re-read the focused
            // field until its text contains the words we injected, which confirms three things at
            // once: the Ctrl+V completed, focus has settled back on the target (not SpeechX's own
            // recording overlay), and we're reading the element we'll go on to watch. A single blind
            // read at a fixed delay is racy — if it fires a beat early, the baseline is missing the
            // injected text and every later diff sees those words as "new", so nothing is ever
            // learned. Best-effort: if the text never matches (e.g. the app doesn't expose it), fall
            // back to the last readable value so behaviour is no worse than before.
            AutomationElement? el = null;
            string? baseline = null;
            var deadline = Environment.TickCount + BaselineTimeoutMs;
            while (true)
            {
                var focused = SafeFocusedElement();
                if (focused != null && !IsOwnProcess(focused)
                    && ResolveTextElement(focused) is ({ } resolved, { } text))
                {
                    el = resolved;
                    baseline = text;
                    if (ContainsAll(words, text)) break;   // injection has landed → trustworthy baseline
                }
                if (Environment.TickCount >= deadline) break;
                Thread.Sleep(BaselinePollMs);
            }

            if (el == null || baseline == null) return;   // never got readable text → nothing to watch

            lock (_gate)
            {
                _element = el;
                _baseline = baseline;
                _lastPolled = baseline;
                _injectedWords = words;
                _lastLearned = null;
                _stopTimer?.Dispose();
                _stopTimer = new Timer(_ => Stop(), null, WatchMs, Timeout.Infinite);
                _pollTimer?.Dispose();
                _pollTimer = new Timer(_ => Poll(), null, PollMs, PollMs);
            }
        }
        catch (Exception e)
        {
            Debug.WriteLine($"[typeover] start failed: {e.Message}");
        }
    }

    /// <summary>
    /// Re-read the focused field on the poll interval. When the text differs from what we last saw,
    /// treat it as an edit and (re)arm the settle timer — we never act on the change directly here.
    /// </summary>
    private void Poll()
    {
        if (Interlocked.Exchange(ref _polling, 1) == 1) return;   // a previous read is still running
        try
        {
            AutomationElement? el;
            string lastPolled;
            lock (_gate)
            {
                el = _element;
                lastPolled = _lastPolled;
            }
            if (el == null) return;   // stopped between ticks

            if (TryGetText(el) is not string current) return;
            if (string.Equals(current, lastPolled, StringComparison.Ordinal)) return;

            bool active;
            lock (_gate)
            {
                active = _element != null;
                if (active) _lastPolled = current;
            }
            if (active) OnEdited();
        }
        catch (Exception e)
        {
            Debug.WriteLine($"[typeover] poll failed: {e.Message}");
        }
        finally
        {
            Interlocked.Exchange(ref _polling, 0);
        }
    }

    /// <summary>
    /// Fires on every detected edit — (re)arm the settle timer so we only evaluate once the user pauses.
    /// Each new keystroke pushes the timer out, so mid-word states like "Aysh" never get learned.
    /// </summary>
    private void OnEdited()
    {
        lock (_gate)
        {
            if (_element == null) return;
            _settleTimer?.Dispose();
            _settleTimer = new Timer(_ => Evaluate(), null, SettleMs, Timeout.Infinite);
        }
    }

    /// <summary>
    /// Run after editing settles: learn the *final* spelling. If we already learned a (now-superseded)
    /// word this session, the corrected value simply supersedes it.
    /// </summary>
    private void Evaluate()
    {
        AutomationElement? el;
        string baseline;
        HashSet<string> injected;
        string? last;
        lock (_gate)
        {
            el = _element;
            baseline = _baseline;
            injected = _injectedWords;
            last = _lastLearned;
        }
        if (el == null) return;

        if (TryGetText(el) is not string current) return;
        if (TypeOverDetector.Correction(injected, baseline, current) is not { } hit) return;
        var (original, corrected) = hit;
        if (string.Equals(corrected, last, StringComparison.OrdinalIgnoreCase)) return;

        lock (_gate) { _lastLearned = corrected; }
        OnCorrection?.Invoke(corrected, original);
    }

    /// <summary>Stop the session: cancel all timers, drop the element, and clear state.</summary>
    public void Stop()
    {
        lock (_gate)
        {
            _pollTimer?.Dispose(); _pollTimer = null;
            _settleTimer?.Dispose(); _settleTimer = null;
            _stopTimer?.Dispose(); _stopTimer = null;
            _element = null;
            _baseline = "";
            _lastPolled = "";
            _injectedWords = new();
            _lastLearned = null;
        }
    }

    /// <summary>
    /// The element we should read text from for a given focus: the focused element itself if it
    /// exposes its text, otherwise a direct child that does. Some apps put keyboard focus on a
    /// container/pane whose editable text actually lives one level down; checking immediate children
    /// (cheap and bounded) picks those up. We deliberately do NOT search the whole descendant subtree
    /// — on Chromium/Electron windows that can be enormous and slow, and those apps don't expose
    /// usable text to UI Automation anyway.
    /// </summary>
    private static (AutomationElement Element, string Text)? ResolveTextElement(AutomationElement focused)
    {
        if (TryGetText(focused) is string direct) return (focused, direct);
        try
        {
            var condition = new OrCondition(
                new PropertyCondition(AutomationElement.IsValuePatternAvailableProperty, true),
                new PropertyCondition(AutomationElement.IsTextPatternAvailableProperty, true));
            var child = focused.FindFirst(TreeScope.Children, condition);
            if (child != null && TryGetText(child) is string childText) return (child, childText);
        }
        catch (Exception e) { Debug.WriteLine($"[typeover] child text search failed: {e.Message}"); }
        return null;
    }

    /// <summary>Current focused element, or null if UIA can't resolve one right now.</summary>
    private static AutomationElement? SafeFocusedElement()
    {
        try { return AutomationElement.FocusedElement; }
        catch (Exception e) { Debug.WriteLine($"[typeover] focused-element read failed: {e.Message}"); return null; }
    }

    /// <summary>True if the element belongs to our own process (e.g. the recording overlay), which we
    /// must not mistake for the user's target field while focus is settling.</summary>
    private static bool IsOwnProcess(AutomationElement el)
    {
        try { return el.Current.ProcessId == Environment.ProcessId; }
        catch { return false; }
    }

    /// <summary>True if every injected word is present in <paramref name="text"/> (case-insensitive) —
    /// our signal that the paste has landed in the field we're looking at.</summary>
    private static bool ContainsAll(HashSet<string> words, string text)
    {
        var have = TypeOverDetector.Tokenize(text).Select(w => w.ToLowerInvariant()).ToHashSet();
        foreach (var w in words)
            if (!have.Contains(w)) return false;
        return true;
    }

    /// <summary>Read the element's text via ValuePattern, falling back to TextPattern. Null if neither.</summary>
    private static string? TryGetText(AutomationElement el)
    {
        try
        {
            if (el.TryGetCurrentPattern(ValuePattern.Pattern, out var vp))
                return ((ValuePattern)vp).Current.Value;
        }
        catch (Exception e) { Debug.WriteLine($"[typeover] value read failed: {e.Message}"); }

        try
        {
            if (el.TryGetCurrentPattern(TextPattern.Pattern, out var tp))
                return ((TextPattern)tp).DocumentRange.GetText(-1);
        }
        catch (Exception e) { Debug.WriteLine($"[typeover] text read failed: {e.Message}"); }

        return null;
    }
}
